### PR TITLE
Support graceful shutdown

### PR DESCRIPTION
**Pull request by maltesander**
# Description

Part of https://github.com/stackabletech/issues/issues/357

- grace period currently 2min

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
# Author
- [x] Changes are OpenShift compatible
- [x] CRD changes approved
- [x] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
- [x] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [x] Code contains useful comments
- [ ] (Integration-)Test cases added
- [x] Documentation added or updated
- [x] Changelog updated
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
```

### CODE REVIEW

File: rust/crd/src/lib.rs
Comment (Techassi) [inline rust/crd/src/lib.rs:1]: Module declarations need to come after imports as per https://doc.rust-lang.org/beta/style-guide/items.html

Also the imports need to use the following order: std, third-party, and finally crate.

File: rust/operator-binary/src/superset_controller.rs
Comment (Techassi) [inline rust/operator-binary/src/superset_controller.rs:1]: Same here.

**Review by Techassi — CHANGES_REQUESTED** [review]

File: rust/operator-binary/src/superset_controller.rs
Comment (maltesander) [reply rust/operator-binary/src/superset_controller.rs:1]: https://github.com/stackabletech/superset-operator/pull/422/commits/f1221c598a9bb54289864c2460d2f2aabcece861

File: rust/crd/src/lib.rs
Comment (maltesander) [reply rust/crd/src/lib.rs:1]: https://github.com/stackabletech/superset-operator/pull/422/commits/f1221c598a9bb54289864c2460d2f2aabcece861

**Review by adwk67 — APPROVED** [review]
LGTM!

File: rust/crd/src/lib.rs [hunk 1]
```diff
@@ -22,26 +20,30 @@ use stackable_operator::{
...
+use std::collections::BTreeMap;
 use strum::{Display, EnumIter, EnumString, IntoEnumIterator};
 
-use crate::{affinity::get_affinity, authentication::SupersetAuthentication};
-
 pub mod affinity;
 pub mod authentication;
 pub mod druidconnection;
 
+use crate::{affinity::get_affinity, authentication::SupersetAuthentication};
```
Comment (Techassi) [inline rust/crd/src/lib.rs:32]: Module definitions should be placed after all imports.

File: rust/operator-binary/src/superset_controller.rs [hunk 2]
```diff
@@ -34,21 +28,33 @@ use stackable_operator::{
...
     SupersetClusterStatus, SupersetConfig, SupersetConfigOptions, SupersetRole, APP_NAME,
-    CONFIG_DIR, LOG_CONFIG_DIR, LOG_DIR, PYTHONPATH, SUPERSET_CONFIG_FILENAME,
+    PYTHONPATH, STACKABLE_CONFIG_DIR, STACKABLE_LOG_CONFIG_DIR, STACKABLE_LOG_DIR,
+    SUPERSET_CONFIG_FILENAME,
+};
+use std::{
+    borrow::Cow,
+    collections::{BTreeMap, HashMap},
+    sync::Arc,
 };
```
Comment (Techassi) [inline rust/operator-binary/src/superset_controller.rs:54]: They should be moved to the top of the file as well.

File: rust/operator-binary/src/superset_controller.rs [hunk 3]
```diff
@@ -34,21 +28,33 @@ use stackable_operator::{
...
 use strum::{EnumDiscriminants, IntoStaticStr};
 
+use crate::operations::graceful_shutdown::add_graceful_shutdown_config;
```
Comment (Techassi) [inline rust/operator-binary/src/superset_controller.rs:57]: Merge imports.

**Review by Techassi — CHANGES_REQUESTED** [review]

File: rust/crd/src/lib.rs [hunk 4]
```diff
@@ -22,26 +20,30 @@ use stackable_operator::{
     role_utils::{GenericRoleConfig, Role, RoleGroup, RoleGroupRef},
     schemars::{self, JsonSchema},
     status::condition::{ClusterCondition, HasStatusCondition},
+    time::Duration,
 };
+use std::collections::BTreeMap;
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/crd/src/lib.rs]: https://github.com/stackabletech/superset-operator/pull/422/commits/6889c0c82d519d2896d3a346ccca117496bbb83e

File: rust/crd/src/lib.rs (hunk 1 above)
Comment (maltesander) [reply rust/crd/src/lib.rs]: https://github.com/stackabletech/superset-operator/pull/422/commits/6889c0c82d519d2896d3a346ccca117496bbb83e

File: rust/operator-binary/src/superset_controller.rs (hunk 2 above)
Comment (maltesander) [reply rust/operator-binary/src/superset_controller.rs]: https://github.com/stackabletech/superset-operator/pull/422/commits/8fef1c56fabb5bf9203737bb488667f18461ed35

File: rust/operator-binary/src/superset_controller.rs (hunk 3 above)
Comment (maltesander) [reply rust/operator-binary/src/superset_controller.rs]: https://github.com/stackabletech/superset-operator/pull/422/commits/8fef1c56fabb5bf9203737bb488667f18461ed35

**Review by Techassi — APPROVED** [review]

**Pull request merged**